File: 2020/Day1/python/day1.py
import os

def read_input(filename):
    return open(filename).readlines()

def get_sum_pair(elements, search) -> tuple:
    size = len(elements)
    for i, element in enumerate(elements):
        print(f"#{i} {element}")
        paired = search - element
        pivot = size//2
        upper_bound = size-1
        lower_bound = i
        stay_in = True
        paired_pos = None
        while stay_in and (upper_bound-lower_bound > 1):
            if elements[pivot] == paired:
                stay_in = False
                paired_pos = (elements[i], elements[pivot])
            elif elements[pivot] < paired:
                lower_bound = pivot
                pivot = (pivot + upper_bound)//2
            else:
                upper_bound = pivot
                pivot = (lower_bound + pivot)//2
            print(f"pivot = {pivot}, lb = {lower_bound}, ub = {upper_bound}")
        if paired_pos is not None:
            break
        
    return paired_pos or (None, None)
    

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description='Process some integers.')
    parser.add_argument('search', metavar='search', type=int, help='The sum integer to find')
    args = parser.parse_args()

    input_lines = read_input(os.path.join(os.path.dirname(__file__),"../input.txt"))
    casted = [ int(x.strip()) for x in input_lines ]
    ordered = sorted(casted)

    num1, num2 = get_sum_pair(ordered, args.search)

    if num1 and num2:
        print("\n\n")
        print(f"Magic Numbers: {num1} + {num2} = {num1+num2}.")
        print(f"The solution is: {num1} x {num2} = {num1*num2}.")
    else:
        print("No valid pair found.")

File: 2020/Day1/golang/day1.go
package main
	
import (
    "bufio"
    "fmt"
    //"io"
    //"io/ioutil"
    "os"
    "strconv"
    "sort"
    "flag"
)

func readLines(path string) ([]int, error) {
    file, err := os.Open(path)
    if err != nil {
        return nil, err
    }
    defer file.Close()

    var lines []int
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        convertedLine, _ := strconv.Atoi(scanner.Text())
        lines = append(lines, convertedLine)
    }
    return lines, scanner.Err()
}

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func getSumPair(elements []int, search int) (int, int) {
    size := len(elements)
    fmt.Printf("size = %d\n", size)
    pairedPosFirst := -1
    pairedPosSecond := -1
    for i, element := range elements {
        fmt.Printf("#%d %d\n", i, element)
        paired := search - element
        pivot := size/2
        upperBound := size-1
        lowerBound := i
        stayIn := true
        for (stayIn && (upperBound-lowerBound > 1)) {
            if (elements[pivot] == paired){
                stayIn = false
                pairedPosFirst = i
                pairedPosSecond = pivot
            } else if (elements[pivot] < paired) {
                lowerBound = pivot
                pivot = (pivot + upperBound)/2
            } else {
                upperBound = pivot
                pivot = (lowerBound + pivot)/2
            }
            fmt.Printf("pivot = %d, lb = %d, ub = %d\n", pivot, lowerBound, upperBound)
        }
        if (pairedPosSecond != -1){ break }
        fmt.Printf("#%d %d\n", i, element)
    }

    if (pairedPosSecond != -1){
        return elements[pairedPosFirst], elements[pairedPosSecond]
    } else {
        return -1, -1
    }
}

func main() {
    var search int
    flag.IntVar(&search ,"search", 2020, "search sum integer.")
    flag.Parse()

	lines, _ := readLines("../input.txt")
    sort.Ints(lines)

    num1, num2 := getSumPair(lines, search)

    if (num1 != -1 && num2 != -1){
        fmt.Print("\n\n")
        fmt.Printf("Magic Numbers: %d + %d = %d.\n", num1, num2, num1+num2)
        fmt.Printf("The solution is: %d x %d = %d.\n", num1, num2, num1*num2)
    } else {
        fmt.Println("No valid Pair found.")
    }


}